domain/resourcegroup/runaway.go

```diff
@@ -75,84 +78,259 @@ type RunawayRecord struct {
+		quarantineChan:        make(chan *QuarantineRecord, maxWatchRecordChannelSize),
+		staleQuarantineRecord: staleQuarantineChan,
+		activeGroup:           make(map[string]int64),
 	}
+	m.insertionCancel = watchList.OnInsertion(func(ctx context.Context, i *ttlcache.Item[string, *QuarantineRecord]) {
+		m.activeLock.Lock()
+		m.activeGroup[i.Value().ResourceGroupName]++
+		m.activeLock.Unlock()
+	})
+	m.evictionCancel = watchList.OnEviction(func(ctx context.Context, er ttlcache.EvictionReason, i *ttlcache.Item[string, *QuarantineRecord]) {
+		m.activeLock.Lock()
+		m.activeGroup[i.Value().ResourceGroupName]--
+		m.activeLock.Unlock()
+		if i.Value().ID == 0 {
+			return
+		}
+		staleQuarantineChan <- i.Value()
+	})
+	return m
 }
 
 // DeriveChecker derives a RunawayChecker from the given resource group
-func (rm *RunawayManager) DeriveChecker(resourceGroupName string, originalSQL string, planDigest string) *RunawayChecker {
+func (rm *RunawayManager) DeriveChecker(resourceGroupName, originalSQL, sqlDigest, planDigest string) *RunawayChecker {
 	group, err := rm.resourceGroupCtl.GetResourceGroup(resourceGroupName)
 	if err != nil || group == nil {
 		logutil.BgLogger().Warn("cannot setup up runaway checker", zap.Error(err))
 		return nil
 	}
-	if group.RunawaySettings == nil {
+	rm.activeLock.RLock()
+	defer rm.activeLock.RUnlock()
+	if group.RunawaySettings == nil && rm.activeGroup[resourceGroupName] == 0 {
 		return nil
 	}
-	return newRunawayChecker(rm, resourceGroupName, group.RunawaySettings, originalSQL, planDigest)
+	return newRunawayChecker(rm, resourceGroupName, group.RunawaySettings, originalSQL, sqlDigest, planDigest)
 }
 
-func (rm *RunawayManager) markQuarantine(resourceGroupName, convict, watchType string, ttl time.Duration, action string, now *time.Time) {
-	key := resourceGroupName + "/" + convict
-	if rm.watchList.Get(key) == nil {
-		rm.queryLock.Lock()
-		if rm.watchList.Get(key) == nil {
-			rm.watchList.Set(key, struct{}{}, ttl)
-		}
-		rm.queryLock.Unlock()
-	}
+func (rm *RunawayManager) markQuarantine(resourceGroupName, convict string, watchType rmpb.RunawayWatchType, action rmpb.RunawayAction, ttl time.Duration, now *time.Time) {
 	var endTime time.Time
-	if ttl == 0 {
-		endTime = longestTime
-	} else {
+	if ttl > 0 {
 		endTime = now.Add(ttl)
 	}
-	select {
-	case rm.quarantineChan <- &QuarantineRecord{
+	record := &QuarantineRecord{
 		ResourceGroupName: resourceGroupName,
 		StartTime:         *now,
 		EndTime:           endTime,
 		Watch:             watchType,
 		WatchText:         convict,
 		Source:            rm.serverID,
-	}:
+		Action:            action,
+	}
+	rm.addWatchList(record, ttl, false)
+	select {
+	case rm.quarantineChan <- record:
 	default:
 		// TODO: add warning for discard flush records
 	}
 }
+
+func (rm *RunawayManager) addWatchList(record *QuarantineRecord, ttl time.Duration, force bool) {
+	key := record.GetRecordKey()
+	item := rm.watchList.Get(key)
+	if force {
+		rm.queryLock.Lock()
```

Comment:
so getting from watchList is lock-free and delete and set isn't?

Reply:
Yes.